plugin/storage/factory.go

```diff
@@ -354,17 +354,13 @@ func (f *Factory) createArchiveSpanWriter() (spanstore.Writer, error) {
 	return factory.CreateSpanWriter()
 }
 
-func (f *Factory) InitArchiveStorage(
-	logger *zap.Logger,
-) (spanstore.Reader, spanstore.Writer) {
+func (f *Factory) InitArchiveStorage() (spanstore.Reader, spanstore.Writer) {
 	reader, err := f.createArchiveSpanReader()
 	if err != nil {
```

Comment:
what is the nature of the errors returned here? Is it "not supported" or a genuine error like "cannot connect"? If it's always the latter we should just return them and let the top caller log them. If it's not-supported then we need to expose that error as a constant and have the caller check for it with `error.Is` to avoid logging the error / stacktrace and only log "archive not supported, not initialized"

Reply:
@yurishkuro It can be both

```
func (f *Factory) createArchiveSpanReader() (spanstore.Reader, error) {
	factory, ok := f.archiveFactories[f.SpanReaderType]
	if !ok {
		return nil, fmt.Errorf("no %s backend registered for span store", f.SpanReaderType)
	}
	return factory.CreateSpanReader()
}

func (f *Factory) createArchiveSpanWriter() (spanstore.Writer, error) {
	factory, ok := f.archiveFactories[f.SpanWriterTypes[0]]
	if !ok {
		return nil, fmt.Errorf("no %s backend registered for span store", f.SpanWriterTypes[0])
	}
	return factory.CreateSpanWriter()
}
```

Reply:
`"no %s backend registered for span store"` is a legit error that should be displayed to the user. I was only thinking of a potential situation where in the past we did an interface cast and if `!ok` the logged "no archive", this should be warning not an error. But in the other PR we had a case where interface cast was actually successful but the implementation was returning a "not implemented" error - that one will by default be logged as an error, but should instead be also logged as a warning.

Reply:
@yurishkuro Do we want that to be printed at an error log? We basically try to initialize the archive storage based on the main span reader type. That error will always be printed when someone doesn't have archive configured. We have this check in `Initialize` that will return an error when someone has archive storage configured but there was an error in initialization
```go
	for kind, factory := range f.archiveFactories {
		if archivable, ok := factory.(storage.ArchiveCapable); ok && archivable.IsArchiveCapable() {
			if err := initializeFactory(kind, factory, "archive"); err != nil {
				return err
			}
		} else {
			delete(f.archiveFactories, kind)
		}
	}
```

Reply:
ideally, if they have archive storage configured but we can't instantiate it we log an error. If they didn't ask for it but we tried anyway and failed then it's a warning (Q: why would we try in this case?).

Reply:
For your first point - we would fail to startup when initialization fails (see https://github.com/jaegertracing/jaeger/blob/main/cmd/all-in-one/main.go#L106-L108). For your second point, we are not really trying per se - the `f.archiveFactories[f.SpanReaderType]` is checking to see if archive storage is configured. If it is, then we try to create it.

Reply:
right, but this issue was started with v1 all-on-one, so it's not just v2 start up. I think v2 model of fungible storage is much better suited to handle this error issue, I am not worried about it. But in v1 we technically also have `archive.enable` flag which is fairly equivalent to v2 way of configuration (the "roles").

Reply:
@yurishkuro Yeah - we do leverage the enable flag for archive storage in `Initialize` so we would fail to start up if archive storage was not configured properly. Do we want different behaviour there?

Reply:
no, failing in such case is correct

Reply:
I think my question still stands - the errors here look like real errors that should be propagated to the caller

Reply:
@yurishkuro Sounds good - do we want to propagate the `no %s backend registered for span store` error as well? I think that one isn't a real error because it would occur whenever someone hasn't configured archive storage.

Reply:
if v2 it's a real error that should fail startup. In v1 I don't know how this error could happen, what does it mean? Most settings are cloned from primary storage, does the error occur when some specific setting is empty for archive but `archive.enbaled=true`?

Reply:
@yurishkuro This flow altogether is only for v1 - v2 wouldn't go through this. It would just initialize the reader and writer directly. For your second question, that is dependent on the storage implementation and someone might run into an error based on the parameters that they've set for their archive storage. I look through the implementation and only [ES](https://github.com/jaegertracing/jaeger/blob/main/plugin/storage/es/factory.go#L197) and [Kafka](https://github.com/jaegertracing/jaeger/blob/main/plugin/storage/kafka/factory.go#L87) return a non-nil error from `CreateSpanReader`.

Reply:
should we return the errors from here?

Reply:
@yurishkuro we could return the errors from here if they come from `CreateSpanReader` and `CreateSpanWriter`

Reply:
yes, why don't we?